crates/core/src/network/connection.rs

```diff
@@ -122,26 +111,148 @@ impl ConnectionSender {
         let res = self
-            .connection
-            .upgrade()
-            .ok_or(NetworkError::ConnectionClosed)?
             .sender
             .send(Message::new(header, body))
             .await
             .map_err(|_| NetworkError::ConnectionClosed);
-        MESSAGE_SENT.increment(1);
-        CONNECTION_SEND_DURATION.record(send_start.elapsed());
+
+        if res.is_ok() {
+            MESSAGE_SENT.increment(1);
+            CONNECTION_SEND_DURATION.record(send_start.elapsed());
+        }
+
+        res
+    }
+
+    fn create_message<M>(
+        &self,
+        message: M,
+        metadata_versions: HeaderMetadataVersions,
+    ) -> Result<(Header, Body), NetworkError>
+    where
+        M: WireEncode + Targeted,
+    {
+        let header = Header::new(
+            metadata_versions[MetadataKind::NodesConfiguration]
+                .expect("nodes configuration version must be set"),
+            metadata_versions[MetadataKind::Logs],
+            metadata_versions[MetadataKind::Schema],
+            metadata_versions[MetadataKind::PartitionTable],
+        );
+        let body =
+            serialize_message(message, self.protocol_version).map_err(ProtocolError::Codec)?;
+        Ok((header, body))
+    }
+
+    /// Tries sending a message on this connection. If there is no capacity, it will fail. Apart
+    /// from this, the method behaves similarly to [`Connection::send`].
+    #[instrument(skip_all, fields(peer_node_id = %self.peer, target_service = ?message.target(), msg = ?message.kind()))]
+    pub fn try_send<M>(
+        &self,
+        message: M,
+        metadata_versions: HeaderMetadataVersions,
+    ) -> Result<(), NetworkError>
+    where
+        M: WireEncode + Targeted,
+    {
+        let send_start = Instant::now();
+        let (header, body) = self.create_message(message, metadata_versions)?;
+        let res = self
+            .sender
+            .try_send(Message::new(header, body))
+            .map_err(|err| match err {
+                TrySendError::Full(_) => NetworkError::Full,
+                TrySendError::Closed(_) => NetworkError::ConnectionClosed,
+            });
+
+        if res.is_ok() {
+            MESSAGE_SENT.increment(1);
+            CONNECTION_SEND_DURATION.record(send_start.elapsed());
+        }
+
         res
     }
 }
 
+struct HeaderMetadataVersions {
+    versions: EnumMap<MetadataKind, Option<Version>>,
+}
+
+impl Index<MetadataKind> for HeaderMetadataVersions {
+    type Output = Option<Version>;
+
+    fn index(&self, index: MetadataKind) -> &Self::Output {
+        &self.versions[index]
+    }
+}
+
+impl PartialEq for Connection {
+    fn eq(&self, other: &Self) -> bool {
+        self.cid == other.cid && self.peer == other.peer
+    }
+}
+
+/// A handle to send messages through a connection. It's safe and cheap to hold
+/// and clone objects of this even if the connection has been dropped.
```

Comment:
The comment is outdated. This is not so cheap anymore since Live<> may hold a lingering cached copy for all metadata types (schema, config, and logs are potentially big). This changes the assumption that this object is cheap to hold even after the connection has been dropped.

Reply:
True. I will update the comment.